mlem/contrib/venv.py

```diff
@@ -0,0 +1,228 @@
+        env_dir = os.path.abspath(self.target)
+
+        self.context = env_spec.ensure_directories(env_dir)
+
+        true_system_site_packages = env_spec.system_site_packages
+        env_spec.system_site_packages = False
+        env_spec.create_configuration(self.context)
+        env_spec.setup_python(self.context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(  # pylint: disable=protected-access
+                self.context
+            )
+        if not env_spec.upgrade:
+            env_spec.setup_scripts(self.context)
+            env_spec.post_setup(self.context)
+        if true_system_site_packages:
+            env_spec.system_site_packages = True
+            env_spec.create_configuration(self.context)
+
+    def get_installed_packages(self, env_dir):
+        if platform.system() == "Windows":
+            env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+        else:
+            env_exe = os.path.join(env_dir, "bin", "python")
+
+        try:
+            return subprocess.check_output([env_exe, "-m", "pip", "freeze"])
+        except FileNotFoundError as e:
+            raise MlemError(f"Executable for pip not found.\n{e}") from e
+        except subprocess.CalledProcessError as e:
+            raise MlemError(
+                f"Couldn't get the list of packages from pip.\n{e}"
+            ) from e
+        except subprocess.TimeoutExpired as e:
+            raise MlemError(
+                f"Determining list of pip packages timed out.\n{e}"
+            ) from e
+
+    def build(self, obj: MlemModel):
+        if self.current_env:
+            if (
+                os.getenv("VIRTUAL_ENV") is None
+                or sys.prefix == sys.base_prefix
+            ):
+                raise MlemError("No virtual environment detected.")
+            echo(EMOJI_PACK + f"Detected the virtual env {sys.prefix}")
+            env_dir = sys.prefix
+            if platform.system() == "Windows":
+                env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+            else:
+                env_exe = os.path.join(env_dir, "bin", "python")
+        else:
+            echo(EMOJI_PACK + f"Creating virtual env {self.target}...")
+            self.create_virtual_env()
+            assert self.context is not None
+            env_dir = self.context.env_dir
+            os.environ["VIRTUAL_ENV"] = self.context.env_dir
+            env_exe = self.context.env_exe
+        echo(EMOJI_PACK + "Installing the required packages...")
+        # Based on recommendation given in https://pip.pypa.io/en/latest/user_guide/#using-pip-from-your-program
+        install_cmd = [env_exe, "-m", "pip", "install"]
+        if self.no_cache:
+            install_cmd.append("--no-cache-dir")
+        install_cmd.extend(obj.requirements.to_pip())
+        try:
+            subprocess.run(install_cmd, check=True)
+        except FileNotFoundError as e:
+            raise MlemError(f"Executable for pip not found.\n{e}") from e
```

Comment:
You still print the actual error in the end and you have multiple lines like this. You can extract error handling in a separate method, or catch all the exceptions at once and raise single exception like `error running pip: {e}`

Reply:
You can also add error msg as a parameter for the method